wurst/event/DamageModify.wurst

```diff
@@ -0,0 +1,462 @@
+
+	// Leech implementation #2
+	function onLeechEffect2()
+		GetTriggerUnit().damageTarget(GetSpellTargetUnit(), LEECH_DAMAGE, ATTACK_TYPE_NORMAL, LEECH_ID)
+
+	init
+		registerSpellEffectEvent(LEECH_ID, function onLeechEffect2)
+		addDamageListener() d ->
+			if d.data == LEECH_ID
+				d.source.addHP(d.getActual())
+				d.source.addEffect(Abilities.vampiricAuraTarget, "origin").destr()
+
+	// Simple damage resistance buff
+	import UniqueDamageListening
+
+	function onBuffEffect()
+		let target = GetSpellTargetUnit()
+		let effect = target.addEffect(Abilities.voodooAuraTarget, "origin")
+		let cb = addTargetDamageModifier() d ->
+			d.amount -= d.initialAmount * .5
+		doAfter(BUFF_DURATION) ->
+			removeTargetDamageModifier(cb)
+			effect.destr()
+
+	init
+		registerSpellEffectEvent(BUFF_ID, function onLeechEffect2)
+*/
+
+// Configuration
+
+/** If Damage instance wasn't created by one of DamageModify package's functions
+	(or `new Damage(...)`), then Damage.data would be set to this value */
+@configurable public constant DAMAGE_NONTRIGGERED_DATA = -1
+
+/** Maximum amount of damage events that can happen in a single event */
+@configurable public constant RECURSION_TRESHOLD = 32
+
+/** Describes default way to create Damage instance if damage event doesn't
+	correspond to any Damage instance. */
+@configurable function getDamageFromEvent() returns Damage
+	return new Damage(GetEventDamageSource(), GetTriggerUnit(), GetEventDamage(), false, false, getDamageType() == DamageType.SPELL ? ATTACK_TYPE_NORMAL : ATTACK_TYPE_CHAOS, null, null, DAMAGE_NONTRIGGERED_DATA, false)
+
+// Wrappers for unit.damageTarget with data attachment
+
+/** Deals triggered damage from unit to target. Additional data can be specified.
+
+	Is a helper for `new Damage(*args*, false)..modify()..deal()`, where last
+	arg specifies whether the instance should not be destroyed during the event.
+	If you need instance after it has been dealt, create it manually. */
+public function unit.damageTarget(unit target, real amount, bool attack, bool ranged, attacktype attackType, damagetype damageType, weapontype weaponType, int data)
+	new Damage(this, target, amount, attack, ranged, attackType, damageType, weaponType, data, false)..deal()
+
+public function unit.damageTarget(unit target, real amount, attacktype attackType, int data)
+	this.damageTarget(target, amount, false, false, attackType, DAMAGE_TYPE_UNIVERSAL, WEAPON_TYPE_WHOKNOWS, data)
+
+public function unit.damageTarget(unit target, real amount, int data)
+	this.damageTarget(target, amount, false, false, ATTACK_TYPE_CHAOS, DAMAGE_TYPE_UNIVERSAL, WEAPON_TYPE_WHOKNOWS, data)
+
+Damage previousDamage = null
+
+/** Returns the damage instance that was created on last Damage event.
+	For both listeners and modifiers it is previous to callback argument.
+
+	Allows to retrieve the instance right after unit.damageTarget() call,
+	but it is not guaranteed. */
+public function getPreviousDamage() returns Damage
```

Comment:
Until Damage was completely processed by the handler, this function will return an instance that was processed before. In case of wrong order recursion (using native damage functions), it would be previous to root Damage, in case of right order recursion, it would be previous to current Damage according to order.

No guarantees due to the recursion possibility - it would return the last, not the first Damage. And if it is used inside damage handler, then it's return value wouldn't change at all from .damageTarget() call, whether that would be native function or wrapper.

Reply:
needs hotdoc adjusments and perhepas same safety?